whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+  [65]byte   signature
+  u32        query_request_len
+  []byte     query_request
+  u8         num_per_chain_responses
+  []byte     per_chain_responses
+  ```
+- On-Chain [WIP] - depends on whether the request is done via VAA or not, this could be chain/emitter/sequence but that wouldn’t work with faster-than-finality
+  ```go
+  u16        sender_chain_id != 0
+  [32]byte   vaa_hash
+  []byte     query_request
+  u8         num_per_chain_responses
+  []byte     per_chain_responses
+  ```
+- Per-Chain Responses
+
+  ```go
+  u16     chain_id
+  u8      type
+  uint32  response_len
+  []byte  response
+  ```
+
+  1. eth_call (query type 1) Response Body
+
+     ```go
+     u64         block_number
+     [32]byte    block_hash
+     u64         block_time_us
+     u8          num_results
+     []byte      results
+     ```
+
+     ```go
+     u32         result_len
+     []byte      result
+     ```
+
+## REST Service
+
+### Request
+
+```ts
+export interface QueryRequest {
+  bytes: string; // As a hex string
+  signatures?: string; // As a hex string
+}
+```
+
+Note that the signature may be omitted if the user is provisioned to allow unsigned requests in the CCQ server configuration. In that case,
+the CCQ server will sign the request.
+
+#### Request Headers
+
+The requests must include the following header.
+
+- "X-API-Key" must match what is configured on the CCQ server.
+
+### Response
+
+- 200 - request reached consensus in < 1m
+
+  - Signatures is the ECDSA signature + a byte for the index of the guardian in the guardian set
+  - Signatures will appear in increasing set index order
+
+  ```ts
+  export interface QueryResponse {
+    bytes: string; // As a hex string
+    signatures?: string; // As a hex string
+  }
+  ```
+
+- 400 - bad request (malformed input or bytes)
+- 403 - [future] e.g. requestor doesn’t have an account set up
+- 500 - [future] failed to reach consensus (e.g. received 14 responses but 7 with one result and 7 with another)
+- 504 - did not reach consensus in < 1m
+
+# Rollout Considerations
+
+Testing against the testnet guardian may not give us an accurate read of gathering consensus or hitting production nodes - there’s only one guardian and it relies on some public and third-party provider nodes.
+
+For demo purposes, a stand alone GCP VM has been set up to host both a guardian and REST server.
+
+# Security Considerations
+
+## DoS Mitigation
+
+The following are being done to reduce the denial of service threat surface:
+
+- CCQ uses a separate P2P channel.
+- The guardians only listen to CCQ requests on that channel, not responses.
+- The guardians do P2P peer ID filtering, so only configured hosts can communicate with the guardians.
+- Only configured wallets can sign requests.
+- Invalid requests are dropped without a gossip response.
+
+Note that to prevent requests from creating undue load on guardians' RPC nodes, a mechanism may be needed to rate-limit or impose service fees upon requestors.
```

Comment:
We really should be cognizant of what makes sense to do in the guardian / ccq query server, and what does not. This is why I suggested changing some of the error codes and designs to match normal web2 stuff.

In web2 land, this is a solved problem that should not be done in the guardian directly. Just wrap it with a reverse proxy like [nginx](https://www.nginx.com/blog/rate-limiting-nginx/) or [haproxy](https://www.haproxy.com/documentation/hapee/latest/load-balancing/traffic-policing/rate-limit-by-url/) and enable lets encrypt tls.

If a dApp wants to hit the ccq query server directly, the dApp will likely be running behind tls and expect this to be running in tls (which it won't). Better to leave some of this functionality to reverse proxies vs adding yet another keypair to the guardian. What do you think?

Reply:
What you described _is_ the motivation of the CCQ server - a proxy service for the (ccq-specific) gossip network.